### PR TITLE
Keep GFM footnote references with preceding code and link spans (#299)

### LINKED ISSUE #299: Reflow divorces GFM footnote reference markers from sentence-ending punctuation
## Description

When `mdtablefix --wrap` reflows a paragraph, it can insert a line break between a sentence-ending punctuation mark and an immediately following GFM footnote reference marker (e.g. `word.[^3]`). The reference marker should be treated as trailing to the preceding word and punctuation cluster and must not be separated from it unless the cost is prohibitive.

## Example

**Before reflow:**

```
<!-- markdownlint-disable MD013 --> Within these comments, a code block is
denoted by triple back-ticks (```). While `rustdoc` defaults to Rust syntax,
explicitly add the `rust` language specifier for clarity.[^3] A doctest
"passes" when it compiles and runs without panicking. To assert specific
outcomes, use the standard macros `assert!`, `assert_eq!`, and
`assert_ne!`.[^3] <!-- markdownlint-enable MD013 -->
```

**After reflow:**

```
<!-- markdownlint-disable MD013 --> Within these comments, a code block is
denoted by triple back-ticks (```). While `rustdoc` defaults to Rust syntax,
explicitly add the `rust` language specifier for clarity.[^3] A doctest
"passes" when it compiles and runs without panicking. To assert specific
outcomes, use the standard macros `assert!`, `assert_eq!`, and `assert_ne!`.
[^3] <!-- markdownlint-enable MD013 -->
```

A line break is inserted between ```assert_ne!`.``` and ```[^3]```, divorcing the reference marker from the sentence it annotates.

## Expected behaviour

A GFM footnote reference marker that immediately follows sentence-ending punctuation (e.g. `word.[^N]`, `word.[^label]`) must be kept on the same line as that punctuation unless doing so would produce a line that significantly exceeds the target width. The marker must travel with the punctuation that precedes it, not be left stranded at the start of the next line.

## Notes

- The inline-wrapping pipeline already promotes `word.[^label]` shapes as grouped spans (PR #280), but this regression demonstrates that the promotion is not being applied universally — at least when the preceding token is itself an atomic span (e.g. an inline code span ending with `!`).
- Related issues: #291, #292, #293, #297.
- Reported in PR #280 by @leynos.

### PR DESCRIPTION
**Pull request by leynos**
## Summary

Closes #299

When `mdtablefix --wrap` reflows a paragraph, GFM footnote reference markers
that immediately follow sentence-ending punctuation on inline code or link
spans were being wrapped onto the next line (for example, `` `assert_ne!`.[^3]
`` becoming `` `assert_ne!`.
[^3] ``). This change extends
[`determine_token_span`](src/wrap/inline.rs) so Code and Link spans absorb
adjacent footnote reference tokens after trailing punctuation has been
coupled, mirroring the existing `promote_footnote_span` behaviour without
changing span kind.

## Changes

- Add a Code/Link branch in the span-grouping loop in
  [`src/wrap/inline.rs`](src/wrap/inline.rs) to keep footnote markers with
  the preceding atomic span.
- Add unit tests in
  [`src/wrap/inline/footnote_tests.rs`](src/wrap/inline/footnote_tests.rs)
  for inline code and link shapes such as `` `assert_ne!`.[^3] `` and
  `[click here](https://example.com).[^1]`.
- Add an integration snapshot in
  [`tests/wrap_unit.rs`](tests/wrap_unit.rs) covering wrap output near a
  line boundary.

## Test plan

- [x] `make check-fmt`
- [x] `make lint`
- [x] `make test`
- [x] CodeRabbit review (`coderabbit review --agent`) — 0 findings

## Summary by Sourcery

Ensure GFM footnote references stay attached to preceding inline code and link spans during wrapping.

Bug Fixes:
- Prevent footnote reference markers after inline code or links from being wrapped onto the next line, keeping them grouped with the preceding span.

Tests:
- Add unit tests for footnote references following inline code and link spans across various punctuation shapes.
- Add an integration snapshot test to verify wrapped output keeps inline code footnote references intact near line boundaries.

### CODE REVIEW

File: src/wrap/inline.rs [hunk 1]
```diff
@@ -314,6 +314,13 @@ pub(super) fn determine_token_span(tokens: &[String], start: usize) -> (usize, u
             continue;
         }
 
+        if matches!(kind, SpanKind::Code | SpanKind::Link) && looks_like_footnote_ref(token) {
+            width += UnicodeWidthStr::width(token.as_str());
+            end += 1;
+            end = extend_punctuation(tokens, end, &mut width);
+            continue;
```
[the remark this answers is not in the document]
Comment (leynos) [reply src/wrap/inline.rs]: @coderabbitai Has this now been resolved in the latest commit?

Use codegraph analysis to determine your answer.

If this comment is now resolved, please mark it as such using the API. Otherwise, please provide an AI agent prompt for the remaining work to be done to address this comment.

**Comment by leynos** [comment]
@coderabbitai Have the following now been resolved?

Annul any requirements that violate the en-GB-oxendict spelling (-ize / -yse / -our) conventions (for example a request to replace "normalize" with "normalise" or "artefact" with "artifact"), or where the requirement unnecessarily increases cyclomatic complexity.

~~~
Please address the comments from this code review:

## Overall Comments
- The new `SpanKind::Code | SpanKind::Link` branch in `determine_token_span` duplicates some of the intent of `promote_footnote_span`; consider extracting a shared helper or reusing the existing logic so future changes to footnote promotion behavior stay consistent across punctuation, code, and link spans.
- In `determine_token_span`, the new code/link + `looks_like_footnote_ref` branch short-circuits before the existing `SpanKind::Link && is_link` branch; it may be worth adding a brief comment explaining this ordering to clarify why link footnote tokens are handled separately from the main link span logic.

## Individual Comments

### Comment 1
<location path="src/wrap/inline/footnote_tests.rs" line_range="30" />
<code_context>
+#[test]
+fn determine_token_span_groups_footnote_reference_after_inline_code() {
</code_context>
<issue_to_address>
**suggestion (testing):** Add a negative test to ensure inline code does *not* absorb a footnote reference when separated by whitespace

These tests cover the grouped/adjacent case. Please also add a complementary test where the footnote is separated by whitespace (e.g. ``"`assert_ne!`. [^3]"`` or ``"`assert_ne!` [^3]"``) to confirm we do *not* group the footnote in that scenario and to codify this boundary condition for `determine_token_span`.
</issue_to_address>

### Comment 2
<location path="src/wrap/inline.rs" line_range="317" />
<code_context>
             continue;
         }

+        if matches!(kind, SpanKind::Code | SpanKind::Link) && looks_like_footnote_ref(token) {
+            width += UnicodeWidthStr::width(token.as_str());
+            end += 1;
</code_context>
<issue_to_address>
**issue (review_instructions):** Refactor the duplicated link/footnote span-handling logic into a shared helper to keep this code DRY while remaining readable.

Extract the common `width`/`end`/`extend_punctuation` logic used here and in the subsequent `SpanKind::Link` branch into a shared helper function or a small local closure. That will avoid duplicating three lines of logic and make future changes to how link/code spans are extended less error-prone while still keeping the code straightforward to read.

<details>
<summary>Review instructions:</summary>

**Path patterns:** `**/*`

**Instructions:**
Keep code DRY, but readable. Use refactoring approaches best suited for the language in question.

</details>
</issue_to_address>
~~~

**Comment by leynos** [comment]
@coderabbitai review
